### PR TITLE
IBM Cloud Deployment and Server (Backend)

### PR DESCRIPTION
**Pull request by ThaiQ**
# Massive Backend Updates
We are now able to deploy both the backend and frontend to their own separate server on the IBM Cloud Foundry.

## Notifications:
Checkout Frontend at: [https://ibm-goodtechscholars.mybluemix.net/](https://ibm-goodtechscholars.mybluemix.net/)
Checkout Backend's status at: [https://ibm-scholars-server.us-south.cf.appdomain.cloud](https://ibm-scholars-server.us-south.cf.appdomain.cloud)

# Changes:
1. Renamed ```api to server```. From now on, backend progress should be done in ```server/server/new folder for your api```. Then add the tables (and their names) to ```server/server/constants.js``` (follow further instructions in there as well).

2. Basic structure is the same:
 - ```npm run server``` : to run server
 - ```npm start``` : to run frontend

3. Added:
- ```npm run deploy-server```  : deploy backend (AKA: the ```server``` folder) to IBM Cloud Foundry
- ```npm run deploy-frontend``` : deploy frontend (AKA: everything but the ```server``` folder) to IBM Cloud Foundry
- ```npm run add-tables``` : initialize new tables to the database
- ```npm run remove-tables``` : remove all tables on the database

*To deploy to the cloud, you will needed to be added as ```dev``` on the project (on the IBM Cloud) by me*
*Pull from ```master``` and ```rebase``` as always*

### CODE REVIEW

File: server/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM registry.access.redhat.com/ubi7/ubi
```
Comment (seemavora) [inline server/Dockerfile:1]: Are we using docker? Why is this file necessary, or is this required to use the cloud?

File: server/LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
+                           Version 2.0, January 2004
```
Comment (seemavora) [inline server/LICENSE:2]: does this file have to be merged in?

File: server/README.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
+---
+
+copyright:
```
Comment (seemavora) [inline server/README.md:3]: same with this file, does this need to be merged in?

File: server/server/constants.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+module.exports = {
+    pool,
+    client,
+    express,
+    app,
+    port,
+    base_URL,
+    tables,
+    tableNames
+  };
```
Comment (seemavora) [inline server/server/constants.js]: empty line?

File: src/const.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+const base_URL = process.env.NODE_ENV === 'production' ?
+'https://ibm-scholars-server.us-south.cf.appdomain.cloud' : 'http://localhost:3001'
+
+module.exports = {
+    base_URL
+  };
```
Comment (seemavora) [inline src/const.js]: empty line

File: server/Dockerfile (hunk 1 above)
Comment (ThaiQ) [reply server/Dockerfile:1]: yes, its docker, I believe the cloud needs it to deploy. also, we can later move this is Kubernetes if needed.

**Review by Annabel1054 — APPROVED** [review]
Looks good !

File: server/LICENSE (hunk 2 above)
Comment (ThaiQ) [reply server/LICENSE:2]: Yes, its best practice for courtesy. Also, its auto-gen from IBM since I used their starter kit, I am keeping it there for acknowledgment. It will also be better for us in the future whenever we are referring back to this project for works or anything. I will edit/move them to the outer folder in the future.

Did you know even SCE has one? (Corev4)
image

File: server/README.md (hunk 3 above)
Comment (ThaiQ) [reply server/README.md:3]: Like mentioned, its copyright from IBM; just leave it there for acknowledgment since we used their starter kit. I am not so sure about how all of this copyright/freeuse policies things will have effects on us.

image

File: server/server/constants.js (hunk 4 above)
Comment (ThaiQ) [reply server/server/constants.js]: thanks, will fix

File: src/const.js (hunk 5 above)
Comment (ThaiQ) [reply src/const.js]: got it

File: server/LICENSE (hunk 2 above)
Comment (nanarboursal) [reply server/LICENSE:2]: there's a lot of sce secrets we don't know about

File: server/cli-config.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# The container name used for the run container
+container-name-run : "eduapp-express-run"
```
Comment (nanarboursal) [inline server/cli-config.yml:2]: is this where we include kubernetes, or have we not used them yet?

File: server/Dockerfile (hunk 1 above)
Comment (ThaiQ) [reply server/Dockerfile:1]: https://cloud.ibm.com/docs/node?topic=node-enable_existing

File: server/cli-config.yml (hunk 6 above)
Comment (ThaiQ) [reply server/cli-config.yml:2]: yeah, its the client config; responsible for configs on the cloud.
That line is docker containers related

File: server/LICENSE (hunk 2 above)
Comment (ThaiQ) [reply server/LICENSE:2]: README is extremely important for documentation, I will update them more at the end

License and conducts are for courtesy purposes

At the end of the day, we want to present this more than just a "student project" when presenting to recruiters or anyone

Comment (seemavora) [reply server/LICENSE:2]: Yes, I have seen the SCE one, but wasn't sure if we needed it for this. Thanks for the explanation. :))

**Review by seemavora — APPROVED** [review]

**Review by nanarboursal — APPROVED** [review]
beautiful, you're amazing thai.

**Comment by surabhigupta2029** [comment]
Looks great thai! Thanks for doing this :)
